# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/PersistentBufferProvider.h
# Commit: f30dd1bddb47
# Full Hash: f30dd1bddb471a4f3487fe0e55e8c4bad0f8962a
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2022-02-12 09:47:43
# Regressor Bug: 1754130
# File Overlap Count: 2
# Description:
#   Bug 1754130 - Support presenting a WebGLFramebuffer to its own swap chain without opaque FB. r=aosmond,jgilbert
#   
#   Most of the support for presenting a WebGLFramebuffer to a swap chain existed as part of the
#   mechanism for opaque WebXR framebuffer support. However, such "opaque" framebuffer are meant
#   to be opaque in the sense that their attachments can't be inspected or changed, which does
# ==============================================================================

diff -r e178aa7d969e -r f30dd1bddb47 gfx/layers/PersistentBufferProvider.h
--- a/gfx/layers/PersistentBufferProvider.h	Fri Feb 11 19:49:55 2022 +0000
+++ b/gfx/layers/PersistentBufferProvider.h	Fri Feb 11 19:49:56 2022 +0000
@@ -10,6 +10,7 @@
 #include "mozilla/Assertions.h"  // for MOZ_ASSERT, etc
 #include "mozilla/RefPtr.h"      // for RefPtr, already_AddRefed, etc
 #include "mozilla/layers/KnowsCompositor.h"
+#include "mozilla/layers/LayersSurfaces.h"
 #include "mozilla/layers/LayersTypes.h"
 #include "mozilla/RefCounted.h"
 #include "mozilla/gfx/Types.h"
@@ -65,13 +66,17 @@
 
   virtual already_AddRefed<gfx::SourceSurface> BorrowSnapshot() = 0;
 
+  /**
+   * Override this if it's possible to read data directly into the DT without
+   * copying to an intermediate snapshot.
+   */
+  virtual bool CopySnapshotTo(gfx::DrawTarget* aDT) { return false; }
+
   virtual void ReturnSnapshot(
       already_AddRefed<gfx::SourceSurface> aSnapshot) = 0;
 
   virtual TextureClient* GetTextureClient() { return nullptr; }
 
-  virtual ClientWebGLContext* AsWebgl() { return nullptr; }
-
   virtual void OnShutdown() {}
 
   virtual bool SetKnowsCompositor(KnowsCompositor* aKnowsCompositor) {
@@ -88,6 +93,13 @@
    * costly (cf. bug 1294351).
    */
   virtual bool PreservesDrawingState() const = 0;
+
+  /**
+   * Provide a WebGL front buffer for compositing, if available.
+   */
+  virtual Maybe<layers::SurfaceDescriptor> GetFrontBuffer() {
+    return Nothing();
+  }
 };
 
 class PersistentBufferProviderBasic : public PersistentBufferProvider {
@@ -133,7 +145,9 @@
 
   bool IsAccelerated() const override { return true; }
 
-  ClientWebGLContext* AsWebgl() override;
+  Maybe<layers::SurfaceDescriptor> GetFrontBuffer() override;
+
+  bool CopySnapshotTo(gfx::DrawTarget* aDT) override;
 
  protected:
   ~PersistentBufferProviderAccelerated() override;